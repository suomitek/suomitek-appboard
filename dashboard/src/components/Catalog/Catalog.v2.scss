.filters-menu {
  padding-left: 1.2rem;
  h5 {
    text-transform: uppercase;
    font-size: small;
    font-weight: 500;
  }
  .filter-section {
    font-weight: 500;
    margin: 0.7rem 0 0.5rem 0;
  }
}

.empty-catalog {
  text-align: center;
  width: 100%;
  height: calc(100vh - 6rem);
  // margin-top: 2rem;
  cds-icon {
    margin-top: 2rem;
    width: 4rem;
    height: 4rem;
  }
  cds-button {
    margin-top: 2rem;
  }
  p {
    margin-top: 0.6rem;
  }
}
