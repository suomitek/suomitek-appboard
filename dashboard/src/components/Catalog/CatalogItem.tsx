import * as React from "react";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import helmIcon from "../../icons/helm.svg";
import operatorIcon from "../../icons/operator-framework.svg";
import placeholder from "../../placeholder.png";
import { IRepo, IStoreState } from "../../shared/types";
import * as url from "../../shared/url";
import InfoCard from "../InfoCard";
import "./CatalogItem.css";

export interface ICatalogItem {
  id: string;
  name: string;
  version: string;
  description: string;
  cluster: string;
  namespace: string;
  icon?: string;
}

export interface IChartCatalogItem extends ICatalogItem {
  repo: IRepo;
}

export interface IOperatorCatalogItem extends ICatalogItem {
  csv: string;
}

export interface ICatalogItemProps {
  type: string;
  item: IChartCatalogItem | IOperatorCatalogItem;
}

// 3 lines description max
const MAX_DESC_LENGTH = 90;

function trimDescription(desc: string): string {
  if (desc.length > MAX_DESC_LENGTH) {
    // Trim to the last word under the max length
    return desc.substr(0, desc.lastIndexOf(" ", MAX_DESC_LENGTH)).concat("...");
  }
  return desc;
}

const CatalogItem: React.SFC<ICatalogItemProps> = props => {
  if (props.type === "operator") {
    const item = props.item as IOperatorCatalogItem;
    return OperatorCatalogItem(item);
  } else {
    const item = props.item as IChartCatalogItem;
    return ChartCatalogItem(item);
  }
};

const OperatorCatalogItem: React.SFC<IOperatorCatalogItem> = props => {
  const { icon, name, csv, version, description, cluster, namespace, id } = props;
  const iconSrc = icon || placeholder;
  // Cosmetic change, remove the version from the csv name
  const csvName = props.csv.split(".v")[0];
  const tag1 = <span>{csvName}</span>;
  const link = url.app.operatorInstances.new(cluster, namespace, csv, id);
  const subIcon = operatorIcon;
  const descriptionC = (
    <div className="ListItem__content__description">{trimDescription(description)}</div>
  );
  return (
    <InfoCard
      key={id}
      title={name}
      link={link}
      info={version || "-"}
      icon={iconSrc}
      description={descriptionC}
      tag1Content={tag1}
      tag1Class={""}
      subIcon={subIcon}
    />
  );
};

const ChartCatalogItem: React.SFC<IChartCatalogItem> = props => {
  const { icon, name, repo, version, description, namespace, id } = props;
  const iconSrc = icon || placeholder;
  const cluster = useSelector((state: IStoreState) => state.clusters.currentCluster);
  const tag1 = (
    <Link
      className="ListItem__content__info_tag_link"
      to={url.app.repo(cluster, namespace, repo.name)}
    >
      {repo.name}
    </Link>
  );
  const link = url.app.charts.get(cluster, namespace, name, repo || ({} as IRepo));
  const subIcon = helmIcon;

  const descriptionC = (
    <div className="ListItem__content__description">{trimDescription(description)}</div>
  );
  return (
    <InfoCard
      key={id}
      title={name}
      link={link}
      info={version || "-"}
      icon={iconSrc}
      description={descriptionC}
      tag1Content={tag1}
      tag1Class={repo ? repo.name : ""}
      subIcon={subIcon}
    />
  );
};

export default CatalogItem;
