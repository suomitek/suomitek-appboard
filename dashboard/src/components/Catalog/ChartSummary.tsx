import ChartMaintainers from "components/ChartView/ChartMaintainers";
import React from "react";
import { IChartAttributes, IChartVersion } from "shared/types";

interface IChartSummaryProps {
  version: IChartVersion;
  chartAttrs: IChartAttributes;
}

function isKubernetesCharts(repoURL: string) {
  return (
    repoURL === "https://kubernetes-charts.storage.googleapis.com" ||
    repoURL === "https://kubernetes-charts-incubator.storage.googleapis.com"
  );
}

export default function ChartSummary({ version, chartAttrs }: IChartSummaryProps) {
  return (
    <div className="left-menu">
      {version.attributes.app_version && (
        <section className="left-menu-subsection" aria-labelledby="chartinfo-versions">
          <h5 className="left-menu-subsection-title" id="chartinfo-versions">
            App Version
          </h5>
          <div>{version.attributes.app_version}</div>
        </section>
      )}
      {chartAttrs.home && (
        <section className="left-menu-subsection" aria-labelledby="chartinfo-versions">
          <h5 className="left-menu-subsection-title" id="chartinfo-versions">
            Home
          </h5>
          <div>
            <a href={chartAttrs.home} target="_blank" rel="noopener noreferrer">
              {chartAttrs.home}
            </a>
          </div>
        </section>
      )}
      {chartAttrs.maintainers?.length > 0 && (
        <section className="left-menu-subsection" aria-labelledby="chartinfo-versions">
          <h5 className="left-menu-subsection-title" id="chartinfo-versions">
            Maintainers
          </h5>
          <div>
            <ChartMaintainers
              maintainers={chartAttrs.maintainers}
              githubIDAsNames={isKubernetesCharts(chartAttrs.repo.url)}
            />
          </div>
        </section>
      )}
      {chartAttrs.sources?.length > 0 && (
        <section className="left-menu-subsection" aria-labelledby="chartinfo-versions">
          <h5 className="left-menu-subsection-title" id="chartinfo-versions">
            Related
          </h5>
          <div>
            <ul>
              {chartAttrs.sources.map((s, i) => (
                <li key={i}>
                  <a href={s} target="_blank" rel="noopener noreferrer">
                    {s}
                  </a>
                </li>
              ))}
            </ul>
          </div>
        </section>
      )}
    </div>
  );
}
