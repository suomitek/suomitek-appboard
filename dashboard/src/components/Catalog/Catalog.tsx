import { RouterAction } from "connected-react-router";
import * as React from "react";
import { Link } from "react-router-dom";

import { IFeatureFlags } from "shared/Config";
import { app } from "shared/url";
import { ForbiddenError, IChart, IChartState, IClusterServiceVersion } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import { CardGrid } from "../Card";
import { MessageAlert } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import CatalogItem, {
  ICatalogItemProps,
  IChartCatalogItem,
  IOperatorCatalogItem,
} from "./CatalogItem";

export interface ICatalogProps {
  charts: IChartState;
  repo: string;
  filter: string;
  fetchCharts: (namespace: string, repo: string) => void;
  pushSearchFilter: (filter: string) => RouterAction;
  cluster: string;
  namespace: string;
  kubeappsNamespace: string;
  getCSVs: (namespace: string) => void;
  csvs: IClusterServiceVersion[];
  featureFlags: IFeatureFlags;
}

interface ICatalogState {
  filter: string;
  listCharts: boolean;
  listOperators: boolean;
}

class Catalog extends React.Component<ICatalogProps, ICatalogState> {
  public state: ICatalogState = {
    filter: "",
    listCharts: true,
    listOperators: true,
  };

  public componentDidMount() {
    const { repo, fetchCharts, filter, namespace, getCSVs, featureFlags } = this.props;
    this.setState({ filter });
    fetchCharts(namespace, repo);
    if (featureFlags.operators) {
      getCSVs(namespace);
    }
  }

  public componentDidUpdate(prevProps: ICatalogProps) {
    if (this.props.filter !== prevProps.filter) {
      this.setState({ filter: this.props.filter });
    }
    if (this.props.repo !== prevProps.repo || this.props.namespace !== prevProps.namespace) {
      this.props.fetchCharts(this.props.namespace, this.props.repo);
    }
    if (this.props.namespace !== prevProps.namespace && this.props.featureFlags.operators) {
      this.props.getCSVs(this.props.namespace);
    }
  }

  public render() {
    const {
      charts: {
        isFetching,
        selected: { error },
        items: allItems,
      },
      cluster,
      namespace,
      pushSearchFilter,
      csvs,
    } = this.props;
    const { listCharts, listOperators } = this.state;
    if (error) {
      const isForbidden = error.constructor === ForbiddenError;
      return (
        <MessageAlert
          level={"error"}
          children={
            <div>
              <h5>Unable to fetch catalog</h5>
              There was an error fetching the catalog.
              {isForbidden && " Please choose a namespace above to which you have access."}
            </div>
          }
        />
      );
    }
    if (!isFetching && allItems.length === 0 && csvs.length === 0) {
      return (
        <MessageAlert
          level={"warning"}
          children={
            <div>
              <h5>Charts not found.</h5>
              Manage your Helm chart repositories in Suomitek-appboard by visiting the{" "}
              <Link to={app.config.apprepositories(cluster, namespace)}>
                App repositories configuration
              </Link>{" "}
              page.
            </div>
          }
        />
      );
    }
    const filteredCharts = this.filteredCharts(allItems);
    const filteredCSVs = this.shouldRenderOperators() ? this.filteredCSVs(csvs) : [];
    const catalogItems = this.getCatalogItems(filteredCharts, filteredCSVs);
    const items = catalogItems.map(c => {
      const keyComponent =
        c.type === "operator"
          ? (c.item as IOperatorCatalogItem).csv
          : (c.item as IChartCatalogItem).repo.name;
      return (
        <CatalogItem type={c.type} key={`${c.type}/${keyComponent}/${c.item.name}`} item={c.item} />
      );
    });
    return (
      <section className="Catalog">
        <PageHeader>
          <h1>Catalog</h1>
          <SearchFilter
            className="margin-l-big"
            placeholder="search charts..."
            onChange={this.handleFilterQueryChange}
            value={this.state.filter}
            onSubmit={pushSearchFilter}
          />
        </PageHeader>
        <LoadingWrapper loaded={!isFetching}>
          <div className="row">
            {this.shouldRenderOperators() && (
              <div className="col-2">
                <div className="margin-b-normal">
                  <span>
                    <b>Type:</b>
                  </span>
                </div>
                <div>
                  <label className="checkbox" key="listcharts">
                    <input type="checkbox" checked={listCharts} onChange={this.toggleListCharts} />
                    <span>Charts</span>
                  </label>
                </div>
                <div>
                  <label className="checkbox" key="listoperators">
                    <input
                      type="checkbox"
                      checked={listOperators}
                      onChange={this.toggleListOperators}
                    />
                    <span>Operators</span>
                  </label>
                </div>
              </div>
            )}
            <div className={this.shouldRenderOperators() ? "col-10" : "col-12"}>
              <CardGrid>{items}</CardGrid>
            </div>
          </div>
        </LoadingWrapper>
      </section>
    );
  }

  private filteredCharts = (charts: IChart[]) => {
    const { filter, listCharts } = this.state;
    if (!listCharts) {
      return [];
    }
    return charts.filter(c => new RegExp(escapeRegExp(filter), "i").test(c.id));
  };

  private filteredCSVs(csvs: IClusterServiceVersion[]) {
    const { filter, listOperators } = this.state;
    if (!listOperators) {
      return [];
    }
    return csvs.filter(c => new RegExp(escapeRegExp(filter), "i").test(c.metadata.name));
  }

  private getCatalogItems(charts: IChart[], csvs: IClusterServiceVersion[]): ICatalogItemProps[] {
    let result: ICatalogItemProps[] = [];
    charts.forEach(c => {
      result = result.concat({
        type: "chart",
        item: {
          id: c.id,
          name: c.attributes.name,
          icon: c.attributes.icon ? `api/assetsvc/${c.attributes.icon}` : undefined,
          version: c.relationships.latestChartVersion.data.app_version,
          description: c.attributes.description,
          repo: c.attributes.repo,
          namespace: this.props.namespace,
          cluster: this.props.cluster,
        },
      });
    });
    csvs.forEach(csv => {
      csv.spec.customresourcedefinitions.owned.forEach(crd => {
        result = result.concat({
          type: "operator",
          item: {
            id: crd.name,
            name: crd.displayName,
            icon: `data:${csv.spec.icon[0].mediatype};base64,${csv.spec.icon[0].base64data}`,
            version: crd.version,
            description: crd.description,
            csv: csv.metadata.name,
            namespace: this.props.namespace,
            cluster: this.props.cluster,
          },
        });
      });
    });
    return result.sort((a, b) => (a.item.name.toLowerCase() > b.item.name.toLowerCase() ? 1 : -1));
  }

  private toggleListCharts = () => {
    this.setState({ listCharts: !this.state.listCharts });
  };

  private toggleListOperators = () => {
    this.setState({ listOperators: !this.state.listOperators });
  };

  private handleFilterQueryChange = (filter: string) => {
    this.setState({
      filter,
    });
  };

  // Render Operators only if there are Operators to render and if we are not
  // filtering by repo
  private shouldRenderOperators = () => {
    const { csvs, repo } = this.props;
    return csvs.length > 0 && !repo;
  };
}

export default Catalog;
