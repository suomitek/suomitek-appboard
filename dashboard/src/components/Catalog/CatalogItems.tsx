import React, { useMemo } from "react";
import { getIcon } from "shared/Operators";
import { IChart, IClusterServiceVersion } from "shared/types";
import CatalogItem, { ICatalogItemProps } from "./CatalogItem.v2";

interface ICatalogItemsProps {
  charts: IChart[];
  csvs: IClusterServiceVersion[];
  cluster: string;
  namespace: string;
}

export default function CatalogItems({ charts, csvs, cluster, namespace }: ICatalogItemsProps) {
  const chartItems: ICatalogItemProps[] = useMemo(
    () =>
      charts.map(c => {
        return {
          type: "chart",
          id: `chart/${c.attributes.repo.name}/${c.id}`,
          item: {
            id: c.id,
            name: c.attributes.name,
            icon: c.attributes.icon ? `api/assetsvc/${c.attributes.icon}` : undefined,
            version: c.relationships.latestChartVersion.data.app_version,
            description: c.attributes.description,
            repo: c.attributes.repo,
            cluster,
            namespace,
          },
        };
      }),
    [charts, cluster, namespace],
  );
  const crdItems: ICatalogItemProps[] = useMemo(
    () =>
      csvs
        .map(csv => {
          if (csv.spec.customresourcedefinitions?.owned) {
            return csv.spec.customresourcedefinitions.owned.map(crd => {
              return {
                type: "operator",
                id: `operator/${csv.metadata.name}/${crd.name}`,
                item: {
                  id: crd.name,
                  name: crd.displayName || crd.name,
                  icon: getIcon(csv),
                  version: crd.version,
                  description: crd.description,
                  csv: csv.metadata.name,
                  cluster,
                  namespace,
                },
              };
            });
          } else {
            return [];
          }
        })
        .flat(),
    [csvs, cluster, namespace],
  );

  const sortedItems = chartItems
    .concat(crdItems)
    .sort((a, b) => (a.item.name.toLowerCase() > b.item.name.toLowerCase() ? 1 : -1));

  if (sortedItems.length === 0) {
    return <p>No application matches the current filter.</p>;
  }
  return (
    <>
      {sortedItems.map(i => (
        <CatalogItem key={i.id} {...i} />
      ))}
    </>
  );
}
