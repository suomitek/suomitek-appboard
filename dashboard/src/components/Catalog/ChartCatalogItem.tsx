import * as React from "react";
import { useSelector } from "react-redux";
import { trimDescription } from "shared/utils";
import helmIcon from "../../icons/helm.svg";
import placeholder from "../../placeholder.png";
import { IRepo, IStoreState } from "../../shared/types";
import * as url from "../../shared/url";
import InfoCard from "../InfoCard/InfoCard.v2";
import { IChartCatalogItem } from "./CatalogItem.v2";

export default function ChartCatalogItem(props: IChartCatalogItem) {
  const { icon, name, repo, version, description, namespace, id } = props;
  const iconSrc = icon || placeholder;
  const cluster = useSelector((state: IStoreState) => state.clusters.currentCluster);
  const link = url.app.charts.get(cluster, namespace, name, repo || ({} as IRepo));
  const subIcon = helmIcon;

  return (
    <InfoCard
      key={id}
      title={name}
      link={link}
      info={version || ""}
      icon={iconSrc}
      description={trimDescription(description)}
      tag1Content={<span>{repo.name}</span>}
      subIcon={subIcon}
    />
  );
}
