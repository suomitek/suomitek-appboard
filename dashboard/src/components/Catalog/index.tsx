import * as React from "react";

import { ICatalogProps } from "./Catalog";

const Catalog = React.lazy(() => import("./Catalog"));
const CatalogV2 = React.lazy(() => import("./Catalog.v2"));

interface ICatalogSelectorProps extends ICatalogProps {
  UI: string;
}

const CatalogSelector: React.FC<ICatalogSelectorProps> = props => (
  <React.Suspense fallback={null}>
    {props.UI === "clarity" ? <CatalogV2 {...props} /> : <Catalog {...props} />}
  </React.Suspense>
);

export default CatalogSelector;
