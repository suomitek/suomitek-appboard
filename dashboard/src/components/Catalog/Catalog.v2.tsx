import { RouterAction } from "connected-react-router";
import { flatten, get, intersection, uniq } from "lodash";
import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { CdsIcon } from "../Clarity/clarity";

import FilterGroup from "components/FilterGroup/FilterGroup";
import Alert from "components/js/Alert";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { IFeatureFlags } from "shared/Config";
import { IChartState, IClusterServiceVersion } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import CardGrid from "../Card/CardGrid.v2";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper.v2";
import PageHeader from "../PageHeader/PageHeader.v2";
import SearchFilter from "../SearchFilter/SearchFilter.v2";

import { CdsButton } from "components/Clarity/clarity";
import { app } from "shared/url";
import "./Catalog.v2.css";
import CatalogItems from "./CatalogItems";

function categoryToReadable(category: string) {
  return category === "" ? "Unknown" : category.replace(/([a-z])([A-Z][a-z])/g, "$1 $2").trimLeft();
}

function getOperatorCategories(c: IClusterServiceVersion): string[] {
  return get(c, "metadata.annotations.categories", "")
    .split(",")
    .map((category: string) => categoryToReadable(category));
}

interface ICatalogProps {
  charts: IChartState;
  repo: string;
  filter: string;
  fetchCharts: (namespace: string, repo: string) => void;
  pushSearchFilter: (filter: string) => RouterAction;
  cluster: string;
  namespace: string;
  kubeappsNamespace: string;
  getCSVs: (namespace: string) => void;
  csvs: IClusterServiceVersion[];
  featureFlags: IFeatureFlags;
}

function Catalog(props: ICatalogProps) {
  const {
    charts: {
      isFetching,
      selected: { error },
      items: charts,
    },
    fetchCharts,
    cluster,
    namespace,
    pushSearchFilter,
    getCSVs,
    csvs,
    repo,
    filter: propsFilter,
  } = props;
  const [searchFilter, setSearchFilter] = useState(propsFilter);
  const [typeFilter, setTypeFilter] = useState([] as string[]);
  const [repoFilter, setRepoFilter] = useState([] as string[]);
  const [categoryFilter, setCategoryFilter] = useState([] as string[]);
  const [operatorProviderFilter, setOperatorProviderFilter] = useState([] as string[]);

  const allRepos = uniq(charts.map(c => c.attributes.repo.name));
  const allProviders = uniq(csvs.map(c => c.spec.provider.name));
  const allCategories = uniq(
    charts
      .map(c => categoryToReadable(c.attributes.category))
      .concat(flatten(csvs.map(c => getOperatorCategories(c)))),
  ).sort();

  useEffect(() => {
    fetchCharts(namespace, repo);
    getCSVs(namespace);
  }, [namespace, repo, fetchCharts, getCSVs]);

  useEffect(() => {
    setSearchFilter(propsFilter);
  }, [propsFilter]);

  const filteredCharts = charts
    .filter(() => typeFilter.length === 0 || typeFilter.includes("Charts"))
    .filter(() => operatorProviderFilter.length === 0)
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.id))
    .filter(c => repoFilter.length === 0 || repoFilter.includes(c.attributes.repo.name))
    .filter(
      c =>
        categoryFilter.length === 0 ||
        categoryFilter.includes(categoryToReadable(c.attributes.category)),
    );
  const filteredCSVs = csvs
    .filter(() => typeFilter.length === 0 || typeFilter.includes("Operators"))
    .filter(() => repoFilter.length === 0)
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.metadata.name))
    .filter(
      c =>
        operatorProviderFilter.length === 0 ||
        operatorProviderFilter.includes(c.spec.provider.name),
    )
    .filter(
      c =>
        categoryFilter.length === 0 ||
        intersection(categoryFilter, getOperatorCategories(c)).length,
    );

  return (
    <section>
      <PageHeader>
        <Row>
          <h1>Catalog</h1>
          <SearchFilter
            key="searchFilter"
            placeholder="search charts..."
            onChange={setSearchFilter}
            value={searchFilter}
            onSubmit={pushSearchFilter}
          />
        </Row>
      </PageHeader>
      <LoadingWrapper loaded={!isFetching}>
        {error && (
          <Alert theme="danger">Found en error fetching the catalog: {error.message}</Alert>
        )}
        {charts.length === 0 && csvs.length === 0 ? (
          <div className="empty-catalog">
            <CdsIcon shape="bundle" />
            <p>The current catalog is empty.</p>
            <p>
              Manage your Helm chart repositories in Suomitek-appboard by visiting the App repositories
              configuration page.
            </p>
            <Link to={app.config.apprepositories(cluster, namespace)}>
              <CdsButton>Manage App Repositories</CdsButton>
            </Link>
          </div>
        ) : (
          <Row>
            <Column span={2}>
              <div className="filters-menu">
                <h5>Filters</h5>
                {csvs.length > 0 && (
                  <div className="filter-section">
                    <label>Application Type:</label>
                    <FilterGroup
                      name="apptype"
                      options={["Operators", "Charts"]}
                      onChange={setTypeFilter}
                    />
                  </div>
                )}
                {allCategories.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Category:</label>
                    <FilterGroup
                      name="category"
                      options={allCategories}
                      onChange={setCategoryFilter}
                    />
                  </div>
                )}
                {allRepos.length > 0 && (
                  <div className="filter-section">
                    <label>Application Repository:</label>
                    <FilterGroup name="apprepo" options={allRepos} onChange={setRepoFilter} />
                  </div>
                )}
                {allProviders.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Operator Provider:</label>
                    <FilterGroup
                      name="operator-provider"
                      options={allProviders}
                      onChange={setOperatorProviderFilter}
                    />
                  </div>
                )}
              </div>
            </Column>
            <Column span={10}>
              <CardGrid>
                <CatalogItems
                  charts={filteredCharts}
                  csvs={filteredCSVs}
                  cluster={cluster}
                  namespace={namespace}
                />
              </CardGrid>
            </Column>
          </Row>
        )}
      </LoadingWrapper>
    </section>
  );
}

export default Catalog;
