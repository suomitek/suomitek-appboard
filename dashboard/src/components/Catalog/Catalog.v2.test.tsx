import context from "jest-plugin-context";
import * as React from "react";

import FilterGroup from "components/FilterGroup/FilterGroup";
import InfoCard from "components/InfoCard/InfoCard.v2";
import Alert from "components/js/Alert";
import { act } from "react-dom/test-utils";
import { defaultStore, mountWrapper } from "shared/specs/mountWrapper";
import itBehavesLike from "../../shared/specs";
import { IChart, IChartState, IClusterServiceVersion } from "../../shared/types";
import SearchFilter from "../SearchFilter/SearchFilter.v2";
import Catalog from "./Catalog.v2";

const defaultChartState = {
  isFetching: false,
  selected: {} as IChartState["selected"],
  deployed: {} as IChartState["deployed"],
  items: [],
  updatesInfo: {},
} as IChartState;
const defaultProps = {
  charts: defaultChartState,
  repo: "",
  filter: "",
  fetchCharts: jest.fn(),
  pushSearchFilter: jest.fn(),
  cluster: "default",
  namespace: "kubeapps",
  kubeappsNamespace: "kubeapps",
  csvs: [],
  getCSVs: jest.fn(),
  featureFlags: { operators: false, ui: "hex" },
};
const chartItem = {
  id: "foo",
  attributes: {
    name: "foo",
    description: "",
    category: "",
    repo: { name: "foo", namespace: "chart-namespace" },
  },
  relationships: { latestChartVersion: { data: { app_version: "v1.0.0" } } },
} as IChart;
const chartItem2 = {
  id: "bar",
  attributes: {
    name: "bar",
    description: "",
    category: "Database",
    repo: { name: "bar", namespace: "chart-namespace" },
  },
  relationships: { latestChartVersion: { data: { app_version: "v2.0.0" } } },
} as IChart;
const csv = {
  metadata: {
    name: "test-csv",
  },
  spec: {
    provider: {
      name: "me",
    },
    icon: [{ base64data: "data", mediatype: "img/png" }],
    customresourcedefinitions: {
      owned: [
        {
          name: "foo-cluster",
          displayName: "foo-cluster",
          version: "v1.0.0",
          description: "a meaningful description",
        },
      ],
    },
  },
} as IClusterServiceVersion;
const populatedProps = {
  ...defaultProps,
  csvs: [csv],
  charts: { ...defaultChartState, items: [chartItem, chartItem2] },
};

it("retrieves csvs in the namespace", () => {
  const getCSVs = jest.fn();
  mountWrapper(defaultStore, <Catalog {...populatedProps} getCSVs={getCSVs} />);
  expect(getCSVs).toHaveBeenCalledWith(defaultProps.namespace);
});

it("shows all the elements", () => {
  const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
  expect(wrapper.find(InfoCard)).toHaveLength(3);
});

it("should render a message if there are no elements in the catalog", () => {
  const wrapper = mountWrapper(defaultStore, <Catalog {...defaultProps} />);
  const message = wrapper.find(".empty-catalog");
  expect(message).toExist();
  expect(message).toIncludeText("The current catalog is empty");
});

it("should render an error if it exists", () => {
  const charts = {
    ...defaultChartState,
    selected: {
      error: new Error("Boom!"),
    },
  } as any;
  const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} charts={charts} />);
  const error = wrapper.find(Alert);
  expect(error.prop("theme")).toBe("danger");
  expect(error).toIncludeText("Boom!");
});

context("when fetching apps", () => {
  itBehavesLike("aLoadingComponent", {
    component: Catalog,
    props: { ...defaultProps, charts: { isFetching: true, items: [], selected: {} } },
  });
});

describe("filters by the searched item", () => {
  it("filters using prop", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} filter={"bar"} />);
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("filters modifying the search box", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
    act(() => {
      (wrapper.find(SearchFilter).prop("onChange") as any)("bar");
    });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});

describe("filters by application type", () => {
  it("doesn't show the filter if there are no csvs", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} csvs={[]} />);
    expect(wrapper.find(FilterGroup).findWhere(g => g.prop("name") === "apptype")).not.toExist();
  });

  it("filters only charts", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Charts");
    input.simulate("change", { target: { value: "Charts" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(2);
  });

  it("filters only operators", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Operators");
    input.simulate("change", { target: { value: "Operators" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});

describe("filters by application repository", () => {
  it("doesn't show the filter if there are no apps", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...defaultProps} />);
    expect(wrapper.find(FilterGroup).findWhere(g => g.prop("name") === "apprepo")).not.toExist();
  });

  it("filters by repo", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
    // The repo name is "foo"
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "foo");
    input.simulate("change", { target: { value: "foo" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});

describe("filters by operator provider", () => {
  it("doesn't show the filter if there are no csvs", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...defaultProps} />);
    expect(
      wrapper.find(FilterGroup).findWhere(g => g.prop("name") === "operator-provider"),
    ).not.toExist();
  });

  it("filters by operator provider", () => {
    const csv2 = {
      metadata: {
        name: "csv2",
      },
      spec: {
        ...csv.spec,
        provider: {
          name: "you",
        },
      },
    } as any;
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} csvs={[csv, csv2]} />);
    // The repo name is "foo"
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "you");
    input.simulate("change", { target: { value: "you" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});

describe("filters by category", () => {
  it("renders a Unknown category if not set", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...defaultProps} charts={{ ...defaultChartState, items: [chartItem] }} />,
    );
    expect(wrapper.find("input").findWhere(i => i.prop("value") === "Unknown")).toExist();
  });

  it("filters a category", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog
        {...defaultProps}
        charts={{ ...defaultChartState, items: [chartItem, chartItem2] }}
      />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(2);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Database");
    input.simulate("change", { target: { value: "Database" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("filters an operator category", () => {
    const csvWithCat = {
      ...csv,
      metadata: {
        name: "csv-cat",
        annotations: {
          categories: "E-Learning",
        },
      },
    } as any;
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...defaultProps} csvs={[csv, csvWithCat]} />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(2);

    const input = wrapper.find("input").findWhere(i => i.prop("value") === "E-Learning");
    input.simulate("change", { target: { value: "E-Learning" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("filters operator categories", () => {
    const csvWithCat = {
      ...csv,
      metadata: {
        name: "csv-cat",
        annotations: {
          categories: "DeveloperTools, Infrastructure",
        },
      },
    } as any;
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...defaultProps} csvs={[csv, csvWithCat]} />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(2);

    // Two categories extracted from the same CSV
    expect(wrapper.find("input").findWhere(i => i.prop("value") === "Developer Tools")).toExist();
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Infrastructure");
    input.simulate("change", { target: { value: "Infrastructure" } });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});
