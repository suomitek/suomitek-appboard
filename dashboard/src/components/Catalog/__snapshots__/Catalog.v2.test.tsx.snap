// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when fetching apps loading spinner matches the snapshot 1`] = `
<section>
  <PageHeader>
    <Row
      list={false}
    >
      <h1>
        Catalog
      </h1>
      <SearchFilter
        key="searchFilter"
        onChange={[Function]}
        onSubmit={[MockFunction]}
        placeholder="search charts..."
        value=""
      />
    </Row>
  </PageHeader>
  <LoadingWrapper
    loaded={false}
  >
    <div
      className="empty-catalog"
    >
      <ReactWrapperComponent
        shape="bundle"
      />
      <p>
        The current catalog is empty.
      </p>
      <p>
        Manage your Helm chart repositories in Kubeapps by visiting the App repositories configuration page.
      </p>
      <Link
        to="/c/default/ns/kubeapps/config/repos"
      >
        <ReactWrapperComponent>
          Manage App Repositories
        </ReactWrapperComponent>
      </Link>
    </div>
  </LoadingWrapper>
</section>
`;
